src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1668,12 +1668,72 @@ namespace Microsoft.FSharp.Core
+                    match box func with
+                    | :? Func<'a, 'a, bool> as f -> f
+                    | _ -> raise (Exception "invalid logic")
+
+                static let _func =
+                    match typeof<'a> with
+                    | t when t.Equals typeof<bool>       -> generalize (Func<_,_,_>(fun (a:bool)      b -> a.Equals b))
+                    | t when t.Equals typeof<sbyte>      -> generalize (Func<_,_,_>(fun (a:sbyte)     b -> a.Equals b))
+                    | t when t.Equals typeof<int16>      -> generalize (Func<_,_,_>(fun (a:int16)     b -> a.Equals b))
+                    | t when t.Equals typeof<int32>      -> generalize (Func<_,_,_>(fun (a:int32)     b -> a.Equals b))
+                    | t when t.Equals typeof<int64>      -> generalize (Func<_,_,_>(fun (a:int64)     b -> a.Equals b))
+                    | t when t.Equals typeof<byte>       -> generalize (Func<_,_,_>(fun (a:byte)      b -> a.Equals b))
+                    | t when t.Equals typeof<uint16>     -> generalize (Func<_,_,_>(fun (a:uint16)    b -> a.Equals b))
+                    | t when t.Equals typeof<uint32>     -> generalize (Func<_,_,_>(fun (a:uint32)    b -> a.Equals b))
+                    | t when t.Equals typeof<uint64>     -> generalize (Func<_,_,_>(fun (a:uint64)    b -> a.Equals b))
+                    | t when t.Equals typeof<nativeint>  -> generalize (Func<_,_,_>(fun (a:nativeint) b -> a.Equals b))
+                    | t when t.Equals typeof<unativeint> -> generalize (Func<_,_,_>(fun (a:unativeint)b -> a.Equals b))
+                    | t when t.Equals typeof<char>       -> generalize (Func<_,_,_>(fun (a:char)      b -> a.Equals b))
+                    | t when t.Equals typeof<string>     -> generalize (Func<_,_,_>(fun (a:string)    b -> System.String.Equals(a,b)))
+                    | t when t.Equals typeof<decimal>    -> generalize (Func<_,_,_>(fun (a:decimal)   b -> System.Decimal.op_Equality(a,b)))
+                    | _ -> null
+                            
+                static member Func = _func
+
+            type GenericSpecializeEqualsWithRelation<'relation, 'a>() =
+                static let generalize (func:Func<'aa,'aa,bool>) =
+                    match box func with
+                    | :? Func<'a, 'a, bool> as f -> f
+                    | _ -> raise (Exception "invalid logic")
+
+                static let _func =
+                    let relationBasedFunc =
+                        match typeof<'relation> with
+                        | r when r.Equals typeof<UnknownEquivalenceRelation> -> null
+                        | r when r.Equals typeof<PartialEquivalenceRelation> ->
+                            match typeof<'a> with
+                            | t when t.Equals typeof<float>   -> generalize (Func<_,_,_>(fun (a:float)   b -> a.Equals b))
+                            | t when t.Equals typeof<float32> -> generalize (Func<_,_,_>(fun (a:float32) b -> a.Equals b))
```

Comment:
I feel a little uneasy about using "a.Equals b" as the base case in all of these different type-specific implementations, both here and [here](https://github.com/Microsoft/visualfsharp/pull/513/files#diff-5470ddaa52d99663f8d391a77ae04f0dR1691). For example these implementations must exactly match the bases cases implemented here https://github.com/Microsoft/visualfsharp/pull/513/files#diff-5470ddaa52d99663f8d391a77ae04f0dL1518 and here: https://github.com/Microsoft/visualfsharp/pull/513/files#diff-5470ddaa52d99663f8d391a77ae04f0dL1758

I think I'd feel more comfortable if we textually repeated ``(# "ceq" x y : bool #)`` here.  For example, I currently have to think carefully "does a.Equals(b) really get optimized down to exactly a ceq instruction?".  Because "a" is a value type I'm not actually totally sure it does.  If we just emit a ceq then I don't need to check.